### ISSUE TITLE
(🐞) Narrower overload impl return included in union creates false error

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#12672]
related to: #12434
```py
from typing import overload

@overload
def f(a: int) -> int: ...
@overload
def f(a: str) -> str: ...

def f(a: int | str) -> bool | str: ...  # E: Overloaded function implementation cannot produce return type of signature 1
```

This impl can definitely produce the return type of signature 1.
